### ISSUE TITLE
[NEW] [Запросы] Использование логического "ИЛИ" в условиях соединения

### ISSUE THREAD

**Issue by artbear** [1c-syntax/bsl-language-server#1561]
## Описание проблемы, ошибки, которую надо диагностировать

Не рекомендуется использовать логическое `ИЛИ` в условиях соединения, то есть в секции `ПО` запроса. Это так же может привести к выбору неоптимального плана и медленной работе запроса. Простого универсального способа переписать такой запрос без использования `ИЛИ` не существует. Следует проанализировать решаемую задачу и попытаться найти другой алгоритм ее решения.

В некоторых случаях возможно переписать запрос через `ОБЪЕДИНИТЬ ВСЕ`
Например, вместо
```bsl
ВЫБРАТЬ Реализация.Ссылка
ИЗ Документ.Реализация.Товары КАК Реализация
ВНУТРЕННЕЕ СОЕДИНЕНИЕ Справочник.Товары КАК Товар
  ПО Реализация.Товар = Товар.Ссылка
  И (Товар.Артикул = "001" ИЛИ Товар.Артикул = "002")
```
можно использовать
```bsl
ВЫБРАТЬ Реализация.Ссылка
ИЗ Документ.Реализация.Товары КАК Реализация
ВНУТРЕННЕЕ СОЕДИНЕНИЕ Справочник.Товары КАК Товар
  ПО Реализация.Товар = Товар.Ссылка
  И Товар.Артикул = "001"

ОБЪЕДИНИТЬ ВСЕ

ВЫБРАТЬ Реализация.Ссылка
ИЗ Документ.Реализация.Товары КАК Реализация
ВНУТРЕННЕЕ СОЕДИНЕНИЕ Справочник.Товары КАК Товар
  ПО Реализация.Товар = Товар.Ссылка
  И Товар.Артикул = "002"
```

## Ссылка на источник, подтверждающее нарушение либо обоснование наличия проблемы

- [Стандарт - Эффективные условия запросов, п.2](https://its.1c.ru/db/v8std/content/658/hdoc)
- [использование логического ИЛИ в условиях - Типичные причины неоптимальной работы запросов и методы оптимизации](https://its.1c.ru/db/content/metod8dev/src/developers/scalability/standards/i8105842.htm#or)

### Тип [Статья на русском](https://1c-syntax.github.io/bsl-language-server/contributing/DiagnosticTypeAndSeverity)

* [ ] :ant: Ошибка
* [ ] :cop: Уязвимость
* [ ] :guardsman: Потенциальная уязвимость
* [x] :poop: Качество кода
* [ ] :trollface: Другое

### Важность [Статья на русском](https://1c-syntax.github.io/bsl-language-server/contributing/DiagnosticTypeAndSeverity)

* [ ] :broken_heart: Блокирующая / Blocker
* [ ] :heart: Критическая / Critical
* [x] :yellow_heart: Важная / Major
* [ ] :blue_heart: Незначительная / Minor
* [ ] :green_heart: Информационная / Info
* [ ] :revolving_hearts: Другое

### Тэги [Статья на русском](https://1c-syntax.github.io/bsl-language-server/contributing/DiagnosticTag)

* [ ] `STANDARD` - "Нарушение стандартов 1С"
* [ ] `LOCKINOS` - "Не будет работать в другой ОС"
* [x] `SQL` - "Проблема с запросом"
* [x] `PERFORMANCE` - "Проблема производительности"
* [ ] `BRAINOVERLOAD` - "Непонятный код"
* [ ] `BADPRACTICE` - "Плохая практика программирования"
* [ ] `CLUMSY` - "Излишние действия"
* [ ] `DESIGN` - "Ошибка в проектировании"
* [ ] `SUSPICIOUS` - "Подозрительный код"
* [x] `UNPREDICTABLE` - "Непредсказуемо работающий код"
* [ ] `DEPRECATED` - "Устаревшая функциональность"
* [ ] `ERROR` - "Ошибочная конструкция"
* [ ] `LOCALIZE` - "Проблемы локализации"

### Время на исправление (минут)

15 минут

**Comment by artbear** [comment]
еще соседняя задача по условия соединения https://github.com/1c-syntax/bsl-language-server/issues/1303

**Comment by ZlobinDV** [comment]
Если мне не изменяет память, то "ИЛИ" опасно в том случае, если идет по разным полям запроса (сработает скан), в твоем же примере `(Товар.Артикул = "001" ИЛИ Товар.Артикул = "002")` идет по одному и тому же полю, что вероятнее всего на стороне SQL превратиться в IN, что не поломает скан и сможет использовать индекс. 

Обсуждал уже этот же момент с Олегом в рамках Пулевского плагина)

**Comment by artbear** [comment]
Да, пример я исправил + добавил еще рекомендацию про временные таблицы

**Comment by artbear** [comment]
сделаю этот ПР после завершения работы с https://github.com/1c-syntax/bsl-language-server/pull/1562 - похожие задачи

**Comment by Levon444** [comment]
Мужчины, если разработка диагностики актуальна мог бы взять в работу.

**Comment by nixel2007** [comment]
@Levon444 актуальна :)

**Comment by Levon444** [comment]
Мужчины, немного задержался с диагностикой. подскажите пожалуйста входит ли в рамки данной диагностики наличие оператора ИЛИ не внутри условия соединения, а между условиями

	```
	...ЛЕВОЕ СОЕДИНЕНИЕ Справочник.СерииНоменклатуры КАК СерииНоменклатуры
         		ПО РеализацияТоваровУслугТовары.Серия = СерииНоменклатуры.Ссылка
         			_ИЛИ_ РеализацияТоваровУслугТовары.КодТНВЭД = СерииНоменклатуры.Номер
         			_ИЛИ_ РеализацияТоваровУслугТовары.КлючСвязи = СерииНоменклатуры.НомерКиЗГИСМ...
```

Это конечно люто редкая штука, но всё же уточняю. Если считаете что контролировать не стоит, то в целом диагностика готова, могу с завтрашнего дня отправить пулл реквест
